libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,508 @@
+
+        return false;
+
+    } else if (last_pos_llh.tow == last_vel_ned.tow
+            && (distMod(last_gps_time.tow, last_vel_ned.tow, TOW_MOD) < 10000)
+            && (distMod(last_dops.tow, last_vel_ned.tow, TOW_MOD) < 60000)
+            && (last_vel_ned.tow > last_full_update_tow || last_gps_time.wn > last_full_update_wn)) {
+
+        //We have an aligned VEL and LLH, and a recent DOPS and TIME.
+
+        //
+        // Check Flags for Valid Messages
+        //
+        if (last_gps_time.flags.fix_mode    == 0 ||
+            last_vel_ned.flags.fix_mode     == 0 ||
+            last_pos_llh.flags.fix_mode     == 0 ||
+            last_dops.flags.fix_mode        == 0) {
+
+            Debug("Message Marked as Invalid. NO FIX! Flags: {GPS_TIME: %d, VEL_NED: %d, POS_LLH: %d, DOPS: %d}",
+                   last_gps_time.flags.fix_mode,
+                   last_vel_ned.flags.fix_mode,
+                   last_pos_llh.flags.fix_mode,
+                   last_dops.flags.fix_mode);
+
+            state.status = AP_GPS::NO_FIX;
+            return false;
+        }
+
+        //
+        // Update external time and accuracy state
+        //
+        state.time_week         = last_gps_time.wn;
+        state.time_week_ms      = last_vel_ned.tow;
+        state.hdop              = last_dops.hdop;
+        state.vdop              = last_dops.vdop;
+        state.last_gps_time_ms = now;
+
+        //
+        // Update velocity state
+        //
+        state.velocity[0]       = (float)(last_vel_ned.n * 1.0e-3);
+        state.velocity[1]       = (float)(last_vel_ned.e * 1.0e-3);
+        state.velocity[2]       = (float)(last_vel_ned.d * 1.0e-3);
+
+        float ground_vector_sq = state.velocity[0]*state.velocity[0] + state.velocity[1]*state.velocity[1];
+        state.ground_speed = safe_sqrt(ground_vector_sq);
+
+        state.ground_course = wrap_360(degrees(atan2f(state.velocity[1], state.velocity[0])));
+
+        state.speed_accuracy        = (float) last_vel_ned.h_accuracy * 1.0e-3f; //Not supported as of SBP2.1.6
+        state.horizontal_accuracy   = (float) last_pos_llh.h_accuracy * 1.0e-3f;
+        state.vertical_accuracy     = (float) last_pos_llh.v_accuracy * 1.0e-3f;
+
+        //
+        // Set flags appropriartely
+        //
+        state.have_vertical_velocity   = true;
+        state.have_speed_accuracy      = state.speed_accuracy != 0;
+        state.have_horizontal_accuracy = state.horizontal_accuracy != 0;
+        state.have_vertical_accuracy   = state.vertical_accuracy != 0;
+
+        //
+        // Update position state
+        //
+        state.location.lat      = (int32_t) (last_pos_llh.lat * (double)1e7);
+        state.location.lng      = (int32_t) (last_pos_llh.lon * (double)1e7);
+        state.location.alt      = (int32_t) (last_pos_llh.height * 100);
+        state.num_sats          = last_pos_llh.n_sats;
+
+        switch (last_pos_llh.flags.fix_mode) {
+            case 0:
```

Comment:
Could this be the default case? It's a 3 bit field, and there are a number of cases you haven't covered here, so if we make this the default case then we've handled other data from the driver sanely.